src/run/service-factory/prompts/BasePrompt.ts

```diff
@@ -0,0 +1,64 @@
+import NewServiceSpec from "../../../model/NewServiceSpec.js";
+import { applySelector, get } from "../utils.js";
+import { confirm } from "../../../helpers/user-input.js";
+
+/**
+ * Abstract implementation of a prompt that will ask the user for input.
+ * Implementations of this class should provide a way to get user input.
+ */
+export default abstract class BasePrompt<T> {
+
+    // eslint-disable-next-line no-useless-constructor
+    constructor (
+        /**
+         * The selector to apply the user input to in the newServiceSpec.
+         * This is a dot separated string that will be used to apply the user input
+         * to the newServiceSpec.
+         */
+        public readonly selector: string,
+
+        /**
+         * Predicate to determine if this prompt should be used for the given service spec.
+         */
+        public readonly predicate?: (newServiceSpec: Partial<NewServiceSpec>) => boolean
+    ) {}
+
+    /**
+     * Get the user input and apply it to the newServiceSpec ensuring that the
+     * user has provided a response to the prompt.
+     * @param newServiceSpec being constructed
+     * @returns amended newServiceSpec with the user input
+     */
+    async make (newServiceSpec: Partial<NewServiceSpec>) {
+        const initialValue = get(newServiceSpec, this.selector);
+
+        // When the value is already set, ask the user if they want to change it, otherwise skip the prompt
+        if (typeof initialValue !== "undefined") {
+            if (!(await confirm(`${this.selector} is already set to "${initialValue}". Do you want to change it?`))) {
+                return newServiceSpec;
+            }
+        }
+
+        while (true) {
```

Comment:
I think we could add a number of retries here to prevent infinite loop for better user experience.

Reply:
Good suggestion thanks for the feedback, I'll take a look and make a change 👍🏼